---
title: "2-RNAvelocity data preparation"
#bibliography: bib.bib
link-citations: yes
date: "`r format (Sys.time(), format='%Y-%m-%d %H:%M:%S %z')`"
output: 
  pdf_document:
    pandoc_args: 
      - '../common.yaml'
      - --listings
    includes: 
      in_header: '../preamble.tex'
    toc: true
    number_sections: true
    fig_width: 7
    fig_height: 6
    fig_caption: true
    df_print: tibble
---

```{r include=FALSE}
knitr::opts_chunk$set(message = FALSE, warning = FALSE, echo = FALSE)
```


\newpage

# Description

Single-cell

# Load data and packages 

```{r}
library(Seurat)
library(ggplot2)
library(dplyr)
library(loomR)
library(tidyverse)

load(file = "../../Data/Objects/Combined.integrated.rds")
seurat.combined <-Combined.integrated
```

# Prepare data RNAvelocyto

Prepare individual Seurat objects for each sample. 

```{r}
list.name.so <- unique(seurat.combined$Sample)
list.name.sample <- list.name.so

for (i in 1:length(list.name.so)) {
  so <- seurat.combined[, seurat.combined$Sample == list.name.so[i]]
  assign(paste(list.name.sample[i], "seuratObject", sep = "."), so)
}

list.name.so <- paste(list.name.sample, "seuratObject", sep = ".")

obj.list <- list()
for (name.so in list.name.so) {
  obj.list <- c(obj.list, get(name.so))
}
list.name.so <- sub(" ", "_", list.name.so)
names(obj.list) <- list.name.so
```

## Generate loom files: 

> The intermediate loom files were too big to be uploaded to the platform but they can be produced by the following steps. 

We counted spliced, unspliced and ambiguous transcripts using velocyto command-line tool (http://velocyto.org)[@LaManno2018]. 

For each sample, the following code was used to generate the loom file: 

```{bash eval=FALSE}
velocyto run 	-b "${sampleID}/outs/filtered_feature_bc_matrix/barcodes.tsv.gz" \
 				      -o "outputDir/${sampleID}.loom" \
 				      "${sampleID}/outs/possorted_genome_bam.bam" \
 				      -m mm10_rmsk.gtf \
 				      /GRCm38/genes/genes.gtf
```

- `${sampleID}` is the sample ID. 
- `${sampleID}/outs` is the output directory of CellRanger. 
- `${sampleID}/outs/possorted_genome_bam.bam` is the BAM file generated from CellRanger. 
- `/refdata-cellranger-GRCh38-3.0.0/genes/genes.gtf` is the gene reference used for Cellranger counts. 

## Read loom files and prepare cellnames

```{r include=FALSE}
suppressMessages(library("velocyto.R"))
list.path.loom <- list.dirs("../../Counts/Loom")
```

Read loom files and create loom objects under sample names
```{r}
list.path.loom <- list.path.loom[-1] # remove the first entry which is parent directory

list.name.loom <- basename(list.path.loom)
list.name.loom <- str_replace(list.name.loom, pattern = "-", replacement = "_")
list.path.loom <- list.files(list.path.loom, pattern = "\\.loom$", full.names = TRUE)

for (i in 1:length(list.name.loom)) {
  assign(make.names(list.name.loom[i]), read.loom.matrices(list.path.loom[i]))
}
```
```{r}
list.name.loom <- make.names(list.name.loom)
```

Make cell names consistent in both loom objects and Seurat objects.

```{r}
# A prefix was added to each cell in Seurat objects during merge step. We should also add prefix to each cell. 

sample.names <- str_remove(list.name.loom, pattern = ".loom")

# find the prefix from seurat object. 
prefix <- sapply(paste0(sample.names,".seuratObject"), 
       function(x) unique(matrix(unlist(strsplit(colnames(get(x)), split = "_")), nrow = 2)[2, ]))

# Add prefix to cellnames. 
source("~/Desktop/velocyto/Script/aggregateLoom.R")
for (i in 1:length(list.name.loom)) {
  loom <- get(list.name.loom[i])  
  assign(list.name.loom[i], value = aggregateLoom(loom, Ori.ID = prefix[i], prefix = FALSE))
}

```

## Filter cellnames and feature names in loom with Seurat gene/cell list

As the Seurat object contains only filtered cells and genes, the genes and cells in loom files should be also filtered. 

```{r}
source("~/Desktop/velocyto/Script/filterLoom.R")

ldat.list <- list()

for (name.sample in list.name.sample) {
  obj.name <- paste(name.sample, "seuratObject", sep = ".")
  loom.name <- paste(name.sample, "loom", sep = ".")
  ldat.name <- paste(name.sample, "ldat", "filtered", sep = ".")
  
  assign(ldat.name, 
         value = filterLoom(loomObj = get(loom.name),
                                geneList = rownames(obj.list[[obj.name]]),
                                cellList = colnames(obj.list[[obj.name]])))
}
```

*All the samples have the same unfound genes (34 genes with redundant symbols). Why? They were mapped in different transcriptome build? *

Remove the redundant genes: 

```{r}
for (name.sample in list.name.sample) {
  obj.name <- paste(name.sample, "seuratObject", sep = ".")
  loom.name <- paste(name.sample, "loom", sep = ".")
  ldat.name <- paste(name.sample, "ldat", "filtered", sep = ".")
  genes.toRemove <- get(paste(name.sample, "ldat.filtered", sep = "."))
    
  genes <- rownames(obj.list[[obj.name]])
  genes.new <- genes[-which(genes %in% genes.toRemove)]
  
  assign(ldat.name, 
         value = filterLoom(loomObj = get(loom.name),
                                geneList = genes.new,
                                cellList = colnames(obj.list[[obj.name]]) ))

  }
```

*ALL FINE!*

Make list of Seurat objects and ldat objects, each under ther sample same.

```{r}
obj <- list()

for (name.sample in list.name.sample) {
  obj.name <- paste(name.sample, "seuratObject", sep = ".")
  ldat.name <- paste(name.sample, "ldat", "filtered", sep = ".")
  
  tmp <- list(ldat = get(ldat.name), seurat = get(obj.name) )
  
  obj[[name.sample]] <- tmp
  }
```

Save for other analyses.

```{r eval=FALSE}
saveRDS(obj, file = "./obj.list.loom_surat.Rds")
```


## Group loom/Seurat objects by treatment

Merge all Seurat objects to one, with only filtered cells. 
Merge all ldat objects to one, with only filtered cells. 

```{r}
obj.all <- obj
# now create merged seurat object and loom data. 

# 1. merged seurat object. 
list.name.sample <- names(obj.all)

seurat.all <- list()
ldat.all <- list()

for (sample.name in list.name.sample) {
  obj <- obj.all[[sample.name]]  
  
  seurat.all[[sample.name]] <- obj[["seurat"]]
  ldat.all[[sample.name]] <- obj[["ldat"]]
}

cellnames <- character()
for (sample.name in list.name.sample) {
  obj <- seurat.all[[sample.name]] 
  
  cellnames <- append(cellnames, colnames(obj))
}

seurat.merge <- seurat.combined[ , cellnames]
seurat.merge
```

```{r}
# 2. merged loom data; 
# source("~/Desktop/velocyto/Script/aggregateLoom.R")

i=1
for (sample.name in list.name.sample) {
  
  obj <- ldat.all[[sample.name]] 
  if (i==1) { 
  spliced <- obj$spliced
  unspliced <- obj$unspliced
  ambiguous <- obj$ambiguous
  } else {
    spliced <- cbind(spliced, obj$spliced)
    unspliced <- cbind(unspliced, obj$unspliced) # note: previous code here was wrong. 
    ambiguous <- cbind(ambiguous, obj$ambiguous) # note: previous code here was wrong. 
  }
    
  i=1+1
}

ldat.merge <- list(spliced=spliced, 
                   unspliced=unspliced, 
                   ambiguous=ambiguous)
```

Now separate them by group. 

```{r}
groupBy <- "Sample"
sample.groupBy <- unique(seurat.merge@meta.data[[groupBy]])

obj <- list()
for (sample.name in sample.groupBy) {
  seurat <- seurat.merge[ , seurat.merge@meta.data[[groupBy]] == sample.name]
  cellnames <- colnames(seurat)
  ldat <- list(spliced = ldat.merge$spliced[, cellnames], 
               unspliced = ldat.merge$unspliced[, cellnames],
               ambiguous = ldat.merge$ambiguous[, cellnames])
  obj[[sample.name]] <- list(ldat=ldat, 
                             seurat=seurat)
}
```

OPTIONAL: save data for other presentations

```{r eval=FALSE}
saveRDS(obj, file = "./obj_group_by_group.list.loom_surat.Rds")
```


# scVelo analysis

## Correct NA in Seurat Metadata

```{r, Correct NA in Seurat Metadata, message=FALSE, warning=FALSE}
 obj <- lapply(obj, function(x)
{
  obj <- x[["seurat"]]
  ldat <- x[["ldat"]]
    for(j in 1:ncol(obj@meta.data)){
          if(is.factor(obj@meta.data[,j]) == T){
          obj@meta.data[,j][is.na(obj@meta.data[,j])] <- "N.A"
      }
          if(is.character(obj@meta.data[,j]) == T){
          obj@meta.data[,j][is.na(obj@meta.data[,j])] <- "N.A"
      }
      }
  x[["seurat"]] <- obj
  x[["ldat"]] <- ldat
  
  return(x)
}
  )
```

## Make loom file from Seurat/loom object 

To facilitate the work, we optimized Seurat `Convert` function to merge a Seurat/Loom list to one Loom file, containing the var matrix with spliced, unspliced layers and obs with all embedding, tsne, umap, pca, clustering, etc. 
A Loom file issue from the function above `Convert.seurat_loom` will be saved in the current working folder.

```{r eval=FALSE}
library(loomR)
source("~/Desktop/velocyto/Script/Convert_Seurat_loom.R")

for (sample.name in sample.groupBy) {
  obj.sl <- obj[[sample.name]]
  pfile <- Convert.seurat_loom(from = obj.sl, to = "loom", filename = paste0(sample.name, ".loom") )
  pfile$close_all()
}

```


## scVelo analysis with dynamical model

> For the details in the estimation of single-cell RNA velocity using dynamical model, refer to the original report[@Bergen2020]: 

> Bergen, V., Lange, M., Peidli, S., Wolf, F. A. & Theis, F. J. Generalizing RNA velocity to transient cell states through dynamical modeling. Nat. Biotechnol. (2020) doi:10.1038/s41587-020-0591-3.

The following codes were used to calculate scRNA velocity and presenting with the existing embedding and labels. 

### For "Non-fumer"

```{python eval=FALSE}
# python below
import scvelo as scv
scv.settings.verbosity = 3  # show errors(0), warnings(1), info(2), hints(3)
scv.settings.presenter_view = True  # set max width size for presenter view
scv.set_figure_params('scvelo')  # for beautified visualization

# load data
ldata_basal = scv.read("./No_Fumer.loom")

# Preprocess the Data
scv.pp.filter_and_normalize(ldata_basal, min_shared_counts=20, n_top_genes=2000)
scv.pp.moments(ldata_basal, n_pcs=30, n_neighbors=30)

# Estimate RNA velocity with dynamical model
scv.tl.recover_dynamics(ldata_basal)
scv.tl.velocity(ldata_basal, mode='dynamical')
scv.tl.velocity_graph(ldata_basal)
scv.pl.velocity_embedding_stream(ldata_basal, basis='umap_cell_embeddings', color='seurat_clusters', 
                                 figsize=(10,10), components='1,2', 
                                 palette=["#2E359A", "#FC990E", "#720D0D", "#6E9BD8"], 
                                 linewidth=1.4, 
                                 title="scVelo analysis", save="No_Fumer.png"
                                )
```
![scVelo results for "Non fumer" grouped sample](Results/scvelo_No_Fumer.png)



# Session information 
```{r}
sessionInfo()
```

# (References)

